# resynthesis

Resynthesis is a time-efficient pipeline written in python that interrogates the genotyping data of a segregating population of plants, aiming to the detection of individuals with complementary genotyping profiles.

## Prerequisites

Apart from python, all the other packages are optional for the user to install. These optional packages are required for graphical representation of data generated by the pipeline.

- python (version 3.8.10 or higher)
- pandas (version 0.25.3 or higher)
- matplotlib (version 3.1.2 or higher)
- seaborn (0.11.2)

**_Note for Windows users_**: instead of using the Command Prompt window, a user can also use the "Terminal" tab in the lower left area of the [Rstudio](https://www.rstudio.com/) environment.

1. **Python installation**

   1. **In Linux**
   
      Follow this [tutorial](https://docs.python-guide.org/starting/install3/linux/) for installing Python 3 in Linux.
      
   2. **In Windows**
   
      Follow this [tutorial](https://phoenixnap.com/kb/how-to-install-python-3-windows) for installing Python 3 in Windows 10. 
      
2. **Pandas, matplotlib and seaborn installation**

   Type `pip install pandas` in the Terminal (Linux and Windows)
   
   **to be continued**...probably all these can be installed from the terminal window in Rstudio, for Windows users
   
## Running the pipeline

XXXXX performs two major tasks of analysis: 1) Detection of homozygous regions (Runs Of Homozygosity; ROH) in the parental line of the segregating population, and 2) Pairwise individual comparison in the population for detecting complementary pairs of individuals.

To see all the avaiable options for running the pipeline, type `python3 XXXXXX.py -h`:

```
usage: resynthesis_master_noImpute.py [-h] [--vcf VCF FILE] [--results_dir STR] [--genos TAB FILE] [--markers TAB FILE] [--not_phased] [--scores_file FILE] [--filter_invariable FLOAT] [--hetero FLOAT]

Detection of ROH regions and analysis of F2 genotyping data for detection of highly complementary individuals.

optional arguments:
  -h, --help            show this help message and exit
  --vcf VCF FILE        Name of the VCF file for the line, for which the ROH regions will be detected.
  --results_dir STR     Name of the results directory [Default: ROH_results]
  --genos TAB FILE      Tab-delimited file (.tab) containing the genotyping data of the F2 population. Markers should be in columns and individuals in rows. Genotypes should be in the format "A,B,H,-", where
                        "-" represents missing data. Incomapatible with --vcf. This arguments is used together with --markers
  --markers TAB FILE    A 2-column tab-delimited file with the markers used for the F2 genotyping, in the format of "chromosome<tab>marker name". Incombatible with "--vcf"
  --not_phased          Use this argument if your genotyping data are not phased. [Default: FALSE]
  --scores_file FILE    A tab delimited file containing user-defined scores for the different combinations of genotypes, observed during the comparison of the individuals. Nucleotides ingenotypes should be
                        separated with a "/". e.g.: A/H<tab>0.75. [Default: scores_default.tab]
  --filter_invariable FLOAT
                        Keep individual pairs that have a percentage of AA or BB combinations that is lower than the argument value. [Default: 0.1]
  --hetero FLOAT        Keep individuals that have a heterozygosity lower than the argument value. [Default: 0.5]

```

Pipeline accepts two inputs, both mutually exclusi
